internal/promoter/image/sign.go

```diff
@@ -42,46 +45,93 @@ const (
+
+	maxParallelVerifications = 10
+	maxParallelSignatures    = 10
+	maxParallelCopies        = 10
 )
 
+// FindSingedEdges takes a list of edges and returns a list of
+// those that have a signature attached
+func (di *DefaultPromoterImplementation) FindSingedEdges(
+	edges map[reg.PromotionEdge]interface{},
+) (map[reg.PromotionEdge]interface{}, error) {
+	signedEdges := map[reg.PromotionEdge]interface{}{}
+	// Verify the signatures in the edges we are looking at
+	t := throttler.New(maxParallelVerifications, len(edges))
+
+	l := sync.Mutex{}
+
+	// First, compile a list of signed images (edges)
+	for e := range edges {
+		go func(edge reg.PromotionEdge) {
+			signer := sign.New(sign.Default())
+			isSigned, err := signer.IsImageSigned(edge.SrcReference())
+			if err != nil {
+				t.Done(
+					fmt.Errorf("checking if %s is signed: %w", edge.SrcReference(), err),
+				)
+				return
+			}
+
+			if !isSigned {
+				t.Done(nil)
+				return
+			}
+			logrus.Infof("Image %s, is signed", edge.SrcReference())
+			l.Lock()
+			signedEdges[edge] = nil
+			l.Unlock()
+			t.Done(nil)
+		}(e)
+		if t.Throttle() > 0 {
+			break
+		}
+	}
+	if err := t.Err(); err != nil {
+		return nil, err
+	}
+	logrus.Infof("%d of promotion candidates are signed", len(signedEdges))
+	return signedEdges, nil
+}
+
 // ValidateStagingSignatures checks if edges (images) have a signature
 // applied during its staging run. If they do it verifies them and
 // returns an error if they are not valid.
 func (di *DefaultPromoterImplementation) ValidateStagingSignatures(
 	edges map[reg.PromotionEdge]interface{},
 ) (map[reg.PromotionEdge]interface{}, error) {
 	signer := sign.New(sign.Default())
-	signedEdges := map[reg.PromotionEdge]interface{}{}
-	for edge := range edges {
-		imageRef := fmt.Sprintf(
-			"%s/%s@%s",
-			edge.SrcRegistry.Name,
-			edge.SrcImageTag.Name,
-			edge.Digest,
-		)
-		logrus.Infof("Verifying signatures of image %s", imageRef)
-
-		// If image is not signed, skip
-		isSigned, err := signer.IsImageSigned(imageRef)
-		if err != nil {
-			return nil, errors.Wrapf(err, "checking if %s is signed", imageRef)
-		}
-
-		if !isSigned {
-			logrus.Infof("No signatures found for ref %s, not checking", imageRef)
-			continue
-		}
-
-		signedEdges[edge] = edges[edge]
+	signedEdges, err := di.FindSingedEdges(edges)
+	if err != nil {
+		return nil, errors.Errorf("filtering signed edges")
+	}
 
-		// Check the staged image signatures
-		if _, err := signer.VerifyImage(imageRef); err != nil {
-			return nil, errors.Wrapf(
-				err, "verifying signatures of image %s", imageRef,
-			)
+	// Verify the signatures in the edges we are looking at
+	t := throttler.New(maxParallelVerifications, len(signedEdges))
+	for e := range signedEdges {
+		go func(edge reg.PromotionEdge) {
+			var err error
+			logrus.Infof("Verifying signatures of image %s", edge.SrcReference())
+
+			// Check the staged image signatures
+			if _, err = signer.VerifyImage(edge.SrcReference()); err != nil {
```

Comment:
```suggestion
			if _, err := signer.VerifyImage(edge.SrcReference()); err != nil {
```
allows to remove the `var err error` in L114

Reply:
Thanks, nice catch!